### PR TITLE
Do not enclose string literal within double quotes before conversion …

### LINKED ISSUE #6862: [Protofuzzer] Error converting case literals to u256
## Description

A proto description like so

```
funcs {
  fd_zero {
    num_input_params: 4
    statements {
      var {
      }
      statements {
        switchstmt {
          switch_expr {
          }
          case_stmt {
            case_lit {
              hexval: "333333333333333333333333333333333333333333333333333333333333333333"
            }
            case_block {
            }
          }
          case_stmt {
            case_lit {
              strval: "333333333333333333333333333333333333333333333333333333333333333333"
            }
            case_block {
            }
          }
        }
      }
    }
  }
  fd_one {
    num_input_params: 8
    statements {
    }
  }
  fd_multi {
    num_input_params: 3
    num_output_params: 4
    statements {
    }
  }
}
```

is thought to contain unique case literals even when it doesn't.

This makes the proto converter believe that it is generating valid yul code when it is in fact going to get rejected by the parser here

https://github.com/ethereum/solidity/blob/b716b211ef712440645da5b3918ca8d997736846/libyul/AsmAnalysis.cpp#L470-L473

The root-cause of the bug is that we enclose a string literal within double quotes before converting to u256. So

`"33...33"` (32 threes) AND `0x33...33` (64 threes)

are thought to be unique (the double quote is `0x22`) so the boost mp library thinks it to be "\x2233...33" (clipping off at the 31'st three most likely)

## Environment

- Compiler version: latest develop

### PR DESCRIPTION
**Pull request by bshastry**
closes #6862

### CODE REVIEW

File: test/tools/ossfuzz/protoToYul.cpp [hunk 1]
```diff
@@ -55,36 +55,40 @@ string ProtoConverter::createAlphaNum(string const& _strBytes) const
...
 		}
-		else
-			tmp = "\"" + tmp + "\"";
 		break;
 	case Literal::LITERAL_ONEOF_NOT_SET:
-		tmp = std::to_string(1);
+		caseLiteralValue = std::to_string(1);
 		break;
 	}
+
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp:80]: Please declare `mpCaseLiteralValue` here

Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp:80]: done.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 2]
```diff
@@ -55,36 +55,40 @@ string ProtoConverter::createAlphaNum(string const& _strBytes) const
...
-	std::string tmp;
+	std::string caseLiteralValue;
 	bool isUnique = false;
 	bool isEmptyString = false;
+
 	switch (_x.literal_oneof_case())
 	{
 	case Literal::kIntval:
-		tmp = std::to_string(_x.intval());
+		caseLiteralValue = std::to_string(_x.intval());
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: This logic seems a little complicated to me. Why don't you go straight away to u256 inside the switch cases? Then you would not need the additional `if` statement at the end.

Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: Simplified logic now.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 3]
```diff
@@ -55,36 +55,37 @@ string ProtoConverter::createAlphaNum(string const& _strBytes) const
...
-	std::string tmp;
+	dev::u256 mpCaseLiteralValue;
 	bool isUnique = false;
 	bool isEmptyString = false;
+
 	switch (_x.literal_oneof_case())
 	{
 	case Literal::kIntval:
-		tmp = std::to_string(_x.intval());
+		mpCaseLiteralValue = dev::u256(std::to_string(_x.intval()));
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Is it necessary to convert to string first?

File: test/tools/ossfuzz/protoToYul.cpp [hunk 4]
```diff
@@ -55,36 +55,36 @@ string ProtoConverter::createAlphaNum(string const& _strBytes) const
...
+		string caseLiteralValue = createAlphaNum(_x.strval());
+		if (caseLiteralValue.empty())
+			caseLiteralValue = std::to_string(0);
+		mpCaseLiteralValue = dev::u256(dev::h256(caseLiteralValue, dev::h256::FromBinary, dev::h256::AlignLeft));
 		break;
 	case Literal::LITERAL_ONEOF_NOT_SET:
-		tmp = std::to_string(1);
+		// If the proto generator does not generate a valid Literal
+		// we generate a case 1:
+		mpCaseLiteralValue = dev::u256(std::to_string(1));
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
		mpCaseLiteralValue = 1;
```

File: test/tools/ossfuzz/protoToYul.cpp (hunk 3 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: No, I'll change it.

**Comment by bshastry** [comment]
Found a bug in the PR. We shouldn't generate stuff like this. The main problem seems to be the equivalence of `""` and 0.

```
{
function foo_noreturn_0(x_0)
{
let x_1
switch 1
case "" {}
case "2" {}
case 0 {}
case 1 {}
}
}
```

P.S. Pushed an update, the bug should hopefully be fixed now.

**Review by chriseth — APPROVED** [review]

**Pull request merged**